### PR TITLE
[AMDGPU] Vectorize more 16 bit shuffles

### PR DESCRIPTION
**Pull request by jrbyrnes**
In the case of larger vectors, we should still prefer the vectorized version (i.e. shufflevector vs extract/insert chains).

In arithmetic chains, vectorization results in chains of packed math instructions (as opposed to unpack/repack & scalarized arithmetic): https://godbolt.org/z/c5onaf6G5

In chains with PHIs, vectorization again removes the unnecessary pack / repack code around BBs: https://godbolt.org/z/vz7zYzvhs

### CODE REVIEW

File: llvm/lib/Target/AMDGPU/AMDGPUTargetTransformInfo.cpp [hunk 1]
```diff
@@ -1135,22 +1135,32 @@ InstructionCost GCNTTIImpl::getShuffleCost(TTI::ShuffleKind Kind,
   if (IsExtractSubvector)
     Kind = TTI::SK_PermuteSingleSrc;
 
+  unsigned NumVectorElts = cast<FixedVectorType>(VT)->getNumElements();
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/AMDGPUTargetTransformInfo.cpp:1138]: Shouldn't assume it's a FixedVector. We can do better than assert on scalable vector

File: llvm/lib/Target/AMDGPU/AMDGPUTargetTransformInfo.cpp [hunk 2]
```diff
@@ -1135,22 +1135,32 @@ InstructionCost GCNTTIImpl::getShuffleCost(TTI::ShuffleKind Kind,
...
+
   if (ST->hasVOP3PInsts()) {
-    if (cast<FixedVectorType>(VT)->getNumElements() == 2 &&
+    if (!(cast<FixedVectorType>(VT)->getNumElements() % 2) &&
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/AMDGPUTargetTransformInfo.cpp:1141]: Reuse the same cast value here

File: llvm/lib/Target/AMDGPU/AMDGPUTargetTransformInfo.cpp [hunk 3]
```diff
@@ -1135,22 +1135,32 @@ InstructionCost GCNTTIImpl::getShuffleCost(TTI::ShuffleKind Kind,
...
-      case TTI::SK_PermuteSingleSrc:
-        return 0;
+      case TTI::SK_PermuteSingleSrc: {
+        if (NumVectorElts == 2)
+          return 0;
+        // Larger vector widths may require additional instructions, but are
+        // typically cheaper than scalarized versions.
+        unsigned RequestedElts =
+            count_if(Mask, [](int MaskElt) { return MaskElt != -1; });
+        return RequestedElts / 2 + RequestedElts % 2;
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/AMDGPUTargetTransformInfo.cpp:1156]: This seems more complicated than necessary. Any odd number would just round up 1?

File: llvm/test/Transforms/SLPVectorizer/AMDGPU/add_sub_sat.ll [hunk 4]
```diff
@@ -211,6 +267,18 @@ define <2 x i32> @ssub_sat_v2i32(<2 x i32> %arg0, <2 x i32> %arg1) {
 ; GCN-NEXT:    [[INS_1:%.*]] = insertelement <2 x i32> [[INS_0]], i32 [[ADD_1]], i64 1
 ; GCN-NEXT:    ret <2 x i32> [[INS_1]]
 ;
+; GFX9-LABEL: @ssub_sat_v2i32(
+; GFX9-NEXT:  bb:
+; GFX9-NEXT:    [[ARG0_0:%.*]] = extractelement <2 x i32> [[ARG0:%.*]], i64 0
+; GFX9-NEXT:    [[ARG0_1:%.*]] = extractelement <2 x i32> [[ARG0]], i64 1
+; GFX9-NEXT:    [[ARG1_0:%.*]] = extractelement <2 x i32> [[ARG1:%.*]], i64 0
+; GFX9-NEXT:    [[ARG1_1:%.*]] = extractelement <2 x i32> [[ARG1]], i64 1
+; GFX9-NEXT:    [[ADD_0:%.*]] = call i32 @llvm.ssub.sat.i32(i32 [[ARG0_0]], i32 [[ARG1_0]])
```
Comment (arsenm) [inline llvm/test/Transforms/SLPVectorizer/AMDGPU/add_sub_sat.ll:276]: Why did i32 cases change? The base VOP3P only should change i16 vectors? Gfx940 has some packed 32-bit ops but I'm not sure this cost model was ever updated to account for that

Comment (jrbyrnes) [reply llvm/test/Transforms/SLPVectorizer/AMDGPU/add_sub_sat.ll]: > Why did i32 cases change?

It was because I changed the way checks were generated. Fixed.

>  Gfx940 has some packed 32-bit ops but I'm not sure this cost model was ever updated to account for that

Looks like the cost model has accurate widths and cost for PackedFP32 https://github.com/llvm/llvm-project/blob/7115ed0fff027b65fa76fdfae215ed1382ed1473/llvm/lib/Target/AMDGPU/AMDGPUTargetTransformInfo.cpp#L609

File: llvm/lib/Target/AMDGPU/AMDGPUTargetTransformInfo.cpp [hunk 5]
```diff
@@ -1135,22 +1135,35 @@ InstructionCost GCNTTIImpl::getShuffleCost(TTI::ShuffleKind Kind,
...
         DL.getTypeSizeInBits(VT->getElementType()) == 16) {
       // With op_sel VOP3P instructions freely can access the low half or high
       // half of a register, so any swizzle is free.
 
       switch (Kind) {
       case TTI::SK_Broadcast:
       case TTI::SK_Reverse:
-      case TTI::SK_PermuteSingleSrc:
-        return 0;
+      case TTI::SK_PermuteSingleSrc: {
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/AMDGPUTargetTransformInfo.cpp:1152]: Should we be specifically handling the other ShuffleKinds? I don't understand what's going on below with the IsExtractSubvector override.

File: llvm/lib/Target/AMDGPU/AMDGPUTargetTransformInfo.cpp [hunk 6]
```diff
@@ -1129,31 +1129,55 @@ InstructionCost GCNTTIImpl::getShuffleCost(TTI::ShuffleKind Kind,
...
-      }
+      unsigned NumPerms = alignTo(RequestedElts, 2) / 2;
+      // SK_Broadcast just reuses the same mask
+      unsigned NumPermMasks = Kind == TTI::SK_Broadcast ? 1 : NumPerms;
+      return NumPerms + NumPermMasks;
+    }
+    case TTI::SK_ExtractSubvector:
+    case TTI::SK_InsertSubvector: {
+      if (HasVOP3P && NumVectorElts == 2)
+        return 0;
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/AMDGPUTargetTransformInfo.cpp:1163]: Any even aligned access of any number of elements should be free, not just the 2 sized ones

File: llvm/lib/Target/AMDGPU/AMDGPUTargetTransformInfo.cpp [hunk 7]
```diff
@@ -1129,31 +1129,55 @@ InstructionCost GCNTTIImpl::getShuffleCost(TTI::ShuffleKind Kind,
...
+      // Insert/extract subvectors require only shifts / extract code to get the
+      // relevant bits
+      return alignTo(RequestedElts, 2) / 2;
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/AMDGPUTargetTransformInfo.cpp:1166]: There's only a shift if you need to access the high half of a register

File: llvm/lib/Target/AMDGPU/AMDGPUTargetTransformInfo.cpp (hunk 6 above)
Comment (jrbyrnes) [reply llvm/lib/Target/AMDGPU/AMDGPUTargetTransformInfo.cpp:1163]: I typically see moves / copies in the even aligned case, we should ignore these? https://godbolt.org/z/h99cKKeY3

File: llvm/lib/Target/AMDGPU/AMDGPUTargetTransformInfo.cpp (hunk 7 above)
Comment (jrbyrnes) [reply llvm/lib/Target/AMDGPU/AMDGPUTargetTransformInfo.cpp:1166]: ditto

File: llvm/lib/Target/AMDGPU/AMDGPUTargetTransformInfo.cpp (hunk 6 above)
Comment (arsenm) [reply llvm/lib/Target/AMDGPU/AMDGPUTargetTransformInfo.cpp:1163]: For the purposes of the cost model, copy should be free.

The copies in the first example aren't a product of the shufflevector itself, those are just shuffling the registers around due to the even aligned register requirement and where they started out in the ABI. Even aligned VGPRs was a new requirement in gfx90a. If you switch to gfx908 or earlier you see these copies disappear.

Comment (jrbyrnes) [reply llvm/lib/Target/AMDGPU/AMDGPUTargetTransformInfo.cpp]: I see -- thanks.

**Review by arsenm — APPROVED** [review]

**Comment by nunoplopes** [comment]
Alive2 complains about this patch:
```llvm
define <2 x i16> @uadd_sat_v9i16_combine_vi16(<9 x i16> %arg0, <9 x i16> %arg1) {
bb:
  %arg0.1 = extractelement <9 x i16> undef, i64 7
  %arg0.2 = extractelement <9 x i16> %arg0, i64 8
  %arg1.1 = extractelement <9 x i16> %arg1, i64 7
  %arg1.2 = extractelement <9 x i16> %arg1, i64 8
  %add.1 = uadd_sat i16 %arg0.1, %arg1.1
  %add.2 = uadd_sat i16 %arg0.2, %arg1.2
  %ins.1 = insertelement <2 x i16> undef, i16 %add.1, i64 0
  %ins.2 = insertelement <2 x i16> %ins.1, i16 %add.2, i64 1
  ret <2 x i16> %ins.2
}
=>
define <2 x i16> @uadd_sat_v9i16_combine_vi16(<9 x i16> %arg0, <9 x i16> %arg1) {
bb:
  %#0 = shufflevector <9 x i16> %arg0, <9 x i16> poison, 4294967295, 8
  %#1 = shufflevector <9 x i16> %arg1, <9 x i16> poison, 7, 8
  %#2 = uadd_sat <2 x i16> %#0, %#1
  ret <2 x i16> %#2
}
Transformation doesn't verify! (unsound)
ERROR: Target is more poisonous than source

Example:
<9 x i16> %arg0 = < poison, poison, poison, poison, poison, poison, poison, poison, poison >
<9 x i16> %arg1 = < poison, poison, poison, poison, poison, poison, poison, #x0000 (0)  [based on undef value], poison >

Source:
i16 %arg0.1 = #x0000 (0)        [based on undef value]
i16 %arg0.2 = poison
i16 %arg1.1 = #x0000 (0)
i16 %arg1.2 = poison
i16 %add.1 = #x0000 (0)
i16 %add.2 = poison
<2 x i16> %ins.1 = < #x0000 (0), #x0000 (0) >
<2 x i16> %ins.2 = < #x0000 (0), poison >

Target:
<2 x i16> %#0 = < poison, poison >
<2 x i16> %#1 = < #x0000 (0), poison >
<2 x i16> %#2 = < poison, poison >
Source value: < #x0000 (0), poison >
Target value: < poison, poison >
```

TL;DR: don't match on insertvalue undef, but just on insertvalue poison.

**Comment by arsenm** [comment]
> Alive2 complains about this patch:

This patch can't have broken this as it just changes the cost model, this must have just added test coverage that hits it

**Comment by nunoplopes** [comment]
> > Alive2 complains about this patch:
> 
> This patch can't have broken this as it just changes the cost model, this must have just added test coverage that hits it

Or that, yes. Sorry, I meant there is a new test failing. The test didn't exist before.

**Comment by alexey-bataev** [comment]
> > > Alive2 complains about this patch:
> > 
> > 
> > This patch can't have broken this as it just changes the cost model, this must have just added test coverage that hits it
> 
> Or that, yes. Sorry, I meant there is a new test failing. The test didn't exist before.

Looks like something wrong in SLP vectorizer, I'll double check
